.github/workflows/ci.yml

```diff
@@ -0,0 +1,156 @@
+
+      - name: Prepare C++ environment
+        run: |
+          . $GITHUB_WORKSPACE/spack-src/share/spack/setup-env.sh
+          spack info fenics-dolfinx
+          spack env create cxx dolfinx-src/.github/workflows/spack-config/gh-actions-env-test.yml
+          spack -e cxx develop --path $GITHUB_WORKSPACE/dolfinx-src fenics-dolfinx@main
+          spack -e cxx develop fenics-basix@git.main=main
+          spack -e cxx develop fenics-ufcx@git.main=main
+          spack -e cxx develop py-fenics-basix@git.main=main
+          spack -e cxx develop py-fenics-ffcx@git.main=main
+          spack -e cxx develop py-fenics-ufl@git.main=main
+
+      - name: Build library (C++)
+        run: |
+          . $GITHUB_WORKSPACE/spack-src/share/spack/setup-env.sh
+          spack -e cxx add fenics-dolfinx@main generator=ninja build_type=Developer +petsc +slepc partitioners=parmetis \
+            %petsc +mpi +hypre +mumps +superlu-dist %openmpi %adios2~fortran~libcatalyst %cmake~ncurses
+          spack -e cxx add catch2 py-fenics-ffcx@main
+          spack -e cxx concretize -j 4
+          spack -e cxx install -j 4 -p 2 --use-buildcache auto
+
+      - name: Push C++ packages to cache
+        run: |
+          . $GITHUB_WORKSPACE/spack-src/share/spack/setup-env.sh
+          spack -e cxx buildcache push --base-image ubuntu:24.04 --only dependencies --with-build-dependencies --update-index local-buildcache
+          spack -e cxx buildcache push --base-image ubuntu:24.04 --update-index local-buildcache catch2 ninja
+        if: ${{ !cancelled() }}
+
+      - name: Build and run C++ unit tests
+        run: |
+          . $GITHUB_WORKSPACE/spack-src/share/spack/setup-env.sh
+          spack env activate cxx
+          spack load catch2 cmake ninja py-fenics-ffcx
+          cmake -Werror=dev --warn-uninitialized -G Ninja -DCMAKE_BUILD_TYPE=Developer -B build/test/ -S dolfinx-src/cpp/test/
+          cmake --build build/test
+          cd build/test
+          ctest -V --output-on-failure -R unittests
+          mpirun -n 3 ctest -V --output-on-failure -R unittests
+
+      - name: Build and run C++ demos
+        run: |
+          . $GITHUB_WORKSPACE/spack-src/share/spack/setup-env.sh
+          spack env activate cxx
+          spack load cmake ninja py-fenics-ffcx
+          cmake -Werror=dev --warn-uninitialized -G Ninja -DCMAKE_BUILD_TYPE=Developer -B build/demo/ -S dolfinx-src/cpp/demo/
+          cmake --build build/demo
+          cd build/demo
+          ctest -V -R demo -R serial
+          ctest --output-on-failure -V -R demo -R mpi_2
+
+      - name: Prepare Python environment
+        run: |
+          . $GITHUB_WORKSPACE/spack-src/share/spack/setup-env.sh
+          spack info py-fenics-dolfinx
+          spack env create py dolfinx-src/.github/workflows/spack-config/gh-actions-env-test.yml
+          spack -e py develop --path $GITHUB_WORKSPACE/dolfinx-src fenics-dolfinx@main
+          spack -e py develop fenics-basix@git.main=main
```

Comment:
These should be checked out using the GitHub action at the git ref specified in the yml file.

Reply:
There is an advantage: the GitHub action does very shallow clones and handles authentication even into private forks.